Implement trailing comma for comma-separated lists

mateusrodriguesxyz opened the pull request:
This is a implementation for https://github.com/swiftlang/swift-evolution/pull/2344 gated behind `-enable-experimental-feature TrailingComma`. After feedback from the language steering group I have expanded the original proposal and implementation that I began in https://github.com/apple/swift/pull/71975.

mateusrodriguesxyz committed:
Implement trailing comma for comma-separated lists

xwu commented:
apple/swift-syntax#2689

@swift-ci build toolchain

mateusrodriguesxyz committed:
Remove enum case list trailing comma support

xwu commented:
apple/swift-syntax#2689

@swift-ci build toolchain

ahoppen commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -1914,6 +1914,10 @@ ParserResult<Expr> Parser::parseExprPrimary(Diag<> ID, bool isExprBasic) {
   UnknownCharacter:
     checkForInputIncomplete();
+    if (Context.LangOpts.hasFeature(Feature::TrailingComma) &&
+        Tok.is(tok::eof) && Tok.getText() == ")") {
```
Could you check for `Tok.is(tok::r_paren)` instead of comparing the text?

mateusrodriguesxyz replied:
So... I don't know why but `Tok.is(tok::r_paren)` doesn't work here 🫠

ahoppen replied:
Do you know what the token’s kind is? Otherwise I’ll try and start debugging it because I’m curious now.

LucianoPAlmeida replied:
Also curious, that if we `checkForInputComplete` wouldn't a call to `!isInputIncomplete()` achieve the same as this check `Tok.is(tok::eof) && Tok.getText() == ")"` does? (note condition is different so that is why I think maybe negate it could have same result)

ahoppen commented on lib/Parse/ParseGeneric.cpp:
```diff
@@ -401,6 +405,8 @@ ParserStatus Parser::parseGenericWhereClause(
       break;
     }
     HasNextReq = consumeIf(tok::comma);
+    IsEndOfList = (Context.LangOpts.hasFeature(Feature::TrailingComma) &&
+                   Tok.is(tok::l_brace));
```
Same comment that I have on the swift-syntax PR here: Can we allow trailing comma in inheritance clauses because `associatedtype` declarations don’t have a  clear terminator?

ahoppen commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -1833,6 +1838,74 @@ Parser::parseStmtConditionElement(SmallVectorImpl<StmtConditionElement> &result,
+      // print("body") }
+      return false;
+    }
+    if (Tok.is(tok::oper_binary_spaced)) {
```
I think you also need to check for `oper_binary_unspaced` here.

mateusrodriguesxyz replied:
@ahoppen do you have a valid example that would `oper_binary_unspaced`? I can't figure out.

ahoppen replied:
That should be something like the following (no spaces around `!=`). Or am I missing something?

```swift
if true, { (x: () -> Void) in true }!=nil { print(0) }
```

mateusrodriguesxyz committed:
Update isStartOfConditionalStmtBody to check for oper_binary_unspaced

mateusrodriguesxyz commented:
> @swift-ci build toolchain

https://ci.swift.org/job/swift-PR-toolchain-macos/1317/

xwu commented:
@mateusrodriguesxyz Would it be wise to add tests that exercise the code paths where you've made changes in response to @ahoppen?

xwu committed:
Merge branch 'main' into trailing-comma

xwu commented:
https://github.com/swiftlang/swift-syntax/pull/2689

@swift-ci build toolchain

xwu commented:
Toolchain links:

* [Ubuntu](https://download.swift.org/tmp/pull-request/74522/899/ubuntu2004/PR-ubuntu2004.tar.gz)
* [Windows](https://ci-external.swift.org/job/swift-PR-build-toolchain-windows/5327/artifact/)
* [macOS](https://download.swift.org/tmp/pull-request/74522/1327/xcode/swift-PR-74522-1327-osx.tar.gz)

LucianoPAlmeida commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3290,7 +3295,7 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
   
   leftLoc = consumeToken(leftTok);
-  return parseList(rightTok, leftLoc, rightLoc, /*AllowSepAfterLast=*/false,
+  return parseList(rightTok, leftLoc, rightLoc, /*AllowSepAfterLast=*/Context.LangOpts.hasFeature(Feature::TrailingComma),
```
Nit: Is this line clang-formatted, it looks longer than 80 columns

mateusrodriguesxyz commented:
@hborla as code owner and member of the language steering group do you have advice about how should I proceed now that [SE-0439 has been accepted with modifications](https://forums.swift.org/t/accepted-with-modifications-se-0439-allow-trailing-comma-in-comma-separated-lists/73216/1)? Should I remove all code unrelated with the accepted design or the group wants to keep some of it (e.g: if/guard/while conditions) under an experimental flag? Should I ungate the accepted parts and remove the flag? I would appreciate any direction, thanks!

hborla commented:
@mateusrodriguesxyz Good question, thanks for asking!

> Should I remove all code unrelated with the accepted design or the group wants to keep some of it (e.g: if/guard/while conditions) under an experimental flag? Should I ungate the accepted parts and remove the flag?

Since this proposal doesn't need the experimental feature flag to be promoted to an upcoming feature flag, I recommend un-gating the accepted parts of the design, and leaving the flag in place to gate the parts of the design that were subsetted out. I also recommend splitting up the tests so that the parts of the tests that need the flag are in a separate file that enable the experimental feature, while the bulk of the tests don't enable the flag.

hborla commented on include/swift/Basic/Features.def:
```diff
@@ -397,6 +397,9 @@ EXPERIMENTAL_FEATURE(ReinitializeConsumeInMultiBlockDefer, false)
 EXPERIMENTAL_FEATURE(SE427NoInferenceOnExtension, false)
 
+// Enable trailing comma for comma-separated lists.
+EXPERIMENTAL_FEATURE(TrailingComma, true)
```
Please switch the argument here to `false`; this controls whether the experimental feature is enable-able in production compiler builds, which this experimental feature shouldn't be. You'll need to `// REQUIRE: asserts` in any tests that use the experimental flag, which is fine!

mateusrodriguesxyz committed:
Ungate accepted parts of SE439

mateusrodriguesxyz committed:
Remove flag from trailing comma tests and move experimental parts to separate file

mateusrodriguesxyz committed:
Update other tests to handle trailing comma

mateusrodriguesxyz commented:
@hborla following your recomendations I've ungated the accepted parts and moved the parts of the tests that need the flag  to a diffeent file. I also removed support from `#available` because it really doens't make sense and while running the full suit of parser tests I noticed that it made the diagnostic worse when `*` is missing.

xwu commented:
@mateusrodriguesxyz Would you mind throwing in tests to ensure that `[Int,]` in type position, `@inline(never,)`, and the like remain invalid? And have you been able to look into which other built-in attributes have custom parameter parsing that doesn't work like a comma-separated parameter list to ensure that we're not adding support for trailing commas in those positions?

mateusrodriguesxyz commented:
> And have you been able to look into which other built-in attributes have custom parameter parsing that doesn't work like a comma-separated parameter list to ensure that we're not adding support for trailing commas in those positions?

I will look into this as soon as possible.

ahoppen commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -1522,6 +1522,10 @@ Parser::parseAvailabilitySpecList(SmallVectorImpl<AvailabilitySpec *> &Specs,
       Status.setIsParseError();
     } else if (consumeIf(tok::comma)) {
+      // End of list with a trailing comma.
+      if (Source != AvailabilitySpecSource::Available && Tok.is(tok::r_paren)) {
+        break;
+      }
```
Why do we need to check for `Source != AvailabilitySpecSource::Available` here?

mateusrodriguesxyz replied:
I did this so `#available` won't support trailing comma. It doesn't really make sense and it leads to worse diagnostic when `*` is missing.

ahoppen replied:
Why don’t we need to check for `AvailabilitySpecSource::Unavailable`?

mateusrodriguesxyz replied:
Makes more sense indeed.

ahoppen commented on test/Parse/availability_query_unavailability.swift:
```diff
@@ -80,9 +80,6 @@ if #unavailable(OSX 10.51, iOS 8.0) {
 if #unavailable(OSX 10.51, { // expected-error {{expected platform name}} // expected-error {{expected ')'}} expected-note {{to match this opening '('}}
 }
 
-if #unavailable(OSX 10.51,) { // expected-error {{expected platform name}}
-}
-
```
Would it make sense to keep these deleted tests around to make sure that we now parse them correctly?

mateusrodriguesxyz replied:
I delete here because I added a similar test case to trailing comma test files.

ahoppen replied:
Ah, I missed that. Thanks.

ahoppen reviewed (commented):
Sorry for taking so long to review this. 

I didn’t review the places where trailing comma is now unconditionally enabled vs enabled behind the `TrailingComma` experimental feature because I lost track of which parts of SE-0439 got accepted. I’d like to review that after the proposal has been updated (assuming that it will be).

For the places where trailing commas are currently only allowed behind the experimental feature, could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

mateusrodriguesxyz committed:
Update trailing comma tests with array literal and built-in attributes

mateusrodriguesxyz commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -1926,6 +1926,9 @@ ParserResult<Expr> Parser::parseExprPrimary(Diag<> ID, bool isExprBasic) {
   default:
   UnknownCharacter:
     checkForInputIncomplete();
+    if (Tok.is(tok::eof) && Tok.getText() == ")") {
+      return nullptr;
+    }
```
This is for handling trailing comma in a string interpolation.

ahoppen replied:
Ah, I see. Could you add that as a comment?

mateusrodriguesxyz commented:
@xwu I've updated the tests with built-in attributes. None of them currently support trailing comma but maybe `backDeployed` and `storageRestrictions` should because they take a variadic list of value as the last argument?

mateusrodriguesxyz commented:
> For the places where trailing commas are currently only allowed behind the experimental feature, could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

I have added these tests at the end:
https://github.com/swiftlang/swift/blob/e4ff4846a78633e1361f7f6e735f06ae16771562/test/Parse/trailing-comma.swift#L112-L139

Did you mean this?

mateusrodriguesxyz commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3290,12 +3293,13 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
   
   leftLoc = consumeToken(leftTok);
-  return parseList(rightTok, leftLoc, rightLoc, /*AllowSepAfterLast=*/false,
+  return parseList(rightTok, leftLoc, rightLoc, /*AllowSepAfterLast=*/true,
```
I don't know the `yield` statement but if it's something like `yield(...)` I believe it should. My plan is to go back to the swift-syntax PR to fix any inconsistent once this PR is approved.

ahoppen replied:
`yield` can be followed by an expression list without parentheses. Eg. `yield x, y` is valid. But `yield x, ` should be disallowed.

xwu replied:
The accepted portions of the proposal are exclusively symmetrically delimited comma-separated lists (i.e., those surrounded by `(...)`, `[...]`, `<...>`, etc.), so this would need to be gated behind a flag.

mateusrodriguesxyz replied:
@ahoppen do you have an example for this? I tried the following but it doesn't parse::
```
struct Foo {
  var x: Int
  var y: Int {
    _read { yield x, } // ❌ consecutive statements on a line must be separated by ';' | expected expression
  }
}
```

ahoppen replied:
Ah, sorry. I did misremember how `yield` works. It does require parentheses if multiple expressions should be yielded, so this is not an issue. Apologies for the confusion.

ahoppen commented on test/Parse/trailing-comma.swift:
```diff
@@ -0,0 +1,139 @@
+// RUN: %target-swift-frontend -parse -verify %s -disable-experimental-parser-round-trip
```
Why do we need to add `-disable-experimental-parser-round-trip`? If we do, that indicates that there is a round-trip bug in the new parser that should be fixed.

mateusrodriguesxyz replied:
Honestly I just copied this from another test file 😅 I will remove it!

ahoppen reviewed (commented):
Could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

mateusrodriguesxyz commented:
> Could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

Sorry, I'm not following. Could you please give an example of what you mean?

ahoppen commented:
> Sorry, I'm not following. Could you please give an example of what you mean?

I was stupid. Didn’t realize that trailing-comma.swift vs trailing-comman-experimental.swift is exactly what I wanted. Sorry for the confusion.

mateusrodriguesxyz committed:
Remove disable-experimental-parser-round-trip from trailing comma tests

xwu commented:
> @xwu I've updated the tests with built-in attributes. None of them currently support trailing comma but maybe backDeployed and storageRestrictions should because they take a variadic list of value as the last argument?

IMO, whatever `@backDeployed` is doing `@available` should also, but as this hasn't been worked through, if all the list handling in built-in attributes are custom and your PR never made any changes to the status quo then we can consider them not part of the proposal and leave it alone.

mateusrodriguesxyz commented:
> if all the list handling in built-in attributes are custom

FWIW `backDeployed` actually uses `parseListItem` with `AllowSepAfterLast` set to `false`.

mateusrodriguesxyz committed:
Minor changes for better clarity

xwu commented:
Ship it?

mateusrodriguesxyz commented:
> Ship it?

Hi @xwu,  just one last thing: Should I remove `@attached` trailing support since it's a built-in attribute?

xwu commented:
I'm kind of surprised to learn it was included in the first place, as it's not really a list. Or are you referred to the `named(...)` part?

mateusrodriguesxyz commented:
No, I mean something like `@attached(extension, conformances: P1, P2,)`.

mateusrodriguesxyz committed:
Remove macro role trailing comma support

mikeash reviewed (changes requested):
There are a lot of formatting changes here. Not sure if we want those changes, but at the very least they should be in their own PR to avoid cluttering the functional changes here, and to make it easier to merge/revert/cherry-pick/etc. this work.

xwu commented:
Got it. I'm not certain, so the safer route is not to add trailing comma support.

mikeash commented:
`git-format` is great! Just don't run it on the entire repository. 😄 Clang includes a `clang-format-diff.py` script which makes it easy to run `git-format` on just your changes.

mikeash commented:
I guess `git-format` got a little too eager? The `clang-format-diff.py` script that comes with clang has worked well for me to apply `clang-format` to just my changes.

eeckstein commented:
Yes, please do not mass-reformat existing code. It makes it nearly impossible to merge between branches without creating a large amount of conflicts.

mateusrodriguesxyz commented:
Sorry about the mess. Should I force push to replace the merge commit or make one reverting the formatting?

al45tair commented:
FWIW, using `git clang-format HEAD^n` where `n` is the number of commits you've made is fine, since that will only reformat things you've changed. Reformatting as you've done here also has the downside that you've pulled in far more reviewers than are likely needed for your actual change.

mateusrodriguesxyz committed:
Merge branch 'main' into trailing-comma

mikeash commented:
That's a much nicer looking diff now, thanks!

ahoppen commented:
@swift-ci Please smoke test

evrensen467 commented:
hey, just a novice here, what does it mean for a feature to be experimental, like this one? i can see in the proposal that this was accepted with modifications, what are the modifications? will we be able to have trailing commas for function arguments?

xwu commented:
You can refer to the [Swift Forums](https://forums.swift.org) for information about new language features adopted through the Swift Evolution process. In this case, the acceptance with modifications is documented [here](https://forums.swift.org/t/accepted-with-modifications-se-0439-allow-trailing-comma-in-comma-separated-lists/73216).

This feature isn't experimental, and no flags will be required to use it when it ships. It has not, however, shipped in any officially released version of Swift yet. You can check the release notes of any release to see what features are included in that release. You can try out development snapshots which contain new features not yet released by going to the [Swift.org Install](https://www.swift.org/install/) page.

Experimental flags are a way to allow work-in-progress to be integrated and tested; they can be added and removed arbitrarily and, for end users, should be disregarded. User-facing features that are meant to be made available to end users generally go through the [Swift Evolution](https://www.swift.org/swift-evolution/) process, which is conducted on the Swift Forums.

ahoppen commented:
@mateusrodriguesxyz What’s the state of this feature? I can see that the Swift Evolution proposal [SE-0439](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0439-trailing-comma-lists.md) has been accepted but the features is still guarded behind an experimental feature flag.

xwu commented:
I believe the experimental feature flag should be guarding those portions of the feature that were not accepted and everything else should be already enabled by default. The proposal text itself has not been edited to incorporate the steering group's decision.

mateusrodriguesxyz commented:
> The proposal text itself has not been edited to incorporate the steering group's decision.

Yes, the text is just outdated. I've been meaning to update it but I've been pretty busy the last months.
The accepted parts are already ungated and working without experimental flag in the **nightly-main** build.

ahoppen commented:
Ah, I remember now. Thanks for clarifying, @mateusrodriguesxyz.
